File: dynamic_programming/climbing_stairs.py
# You are climbing a staircase. It takes n steps to reach the top.
# Each time you can either climb 1 or 2 steps. In how many distinct ways can you climb to the top?

def climb(n):
	if n == 1:
		return 1
	first = 1
	second = 2
	for index in range(3, n+1):
		third = first + second
		first = second
		second = third
	return second


# OR DP

def climb_stairs(n):
	if n == 1:
		return 1
	dp = [0] * (n + 1)
	dp[1] = 1
	dp[2] = 2

	for index in range(3, n+1):
		dp[index] = dp[index - 1] + dp[index - 2]

	return dp[n]

File: dynaprog/climbing_stairs.go
package main

import (
	"fmt"
)

// climb stairs, each time you can take 1 or 2 steps. How many distinct ways can you climb to the top?
// Time: O(n)
// Space: O(n)
func climb(stairs int) int {
	var cache = make([]int, stairs+1)
	return climbing(stairs, cache)
}

func climbing(stairs int, cache []int) int {
	if stairs < 3 {
		return stairs
	}

	if cache[stairs] != 0 {
		return cache[stairs]
	}

	result := climbing(stairs - 1, cache) + climbing(stairs - 2, cache)
	cache[stairs] = result
	return result
}

// Bottom up
// Time: O(n)
// Space: O(n)

func climbAgain(n int) int {
	if n < 3 {
		return n
	}

	ways := make([]int, n+1)
	ways[1] = 1
	ways[2] = 2

	for i := 3 ; i <= n ; i++ {
		ways[i] = ways[i-1] + ways[i-2]
	}

	return ways[n]
}

// Bottom up (Optimized)
// Time: O(n)
// Space: O(1)

func climbOptimized(n int) int {
	if n < 3 {
		return n
	}

	prev, current := 1, 2
	for i := 3 ; i <= n ; i++ {
		prev, current = current, prev+current
	}

	return current
}

func main() {
	fmt.Println(climbOptimized(2))
	fmt.Println(climbOptimized(45))
}